include/wil/registry.h

```diff
@@ -17,16 +17,2704 @@
 
 #include <winreg.h>
 #include <new.h> // new(std::nothrow)
-#include "resource.h" // unique_hkey
+#include "registry_helpers.h"
+#include "resource.h"
+
+#ifndef __WIL_WINREG_
+#error __WIL_WINREG_ is required for ::wil::unique_hkey support; ensure the raw registry APIs (like RegCreateKeyExA) are available
+#endif
+
+// wil registry does not require the use of the STL or C++ exceptions (see _nothrow functions)
+// wil registry natively supports std::vector and std::wstring when preferring those types
+// wil registry uses the __WIL_WINREG_STL define to enable support for wil::shared_* types (defined in resource.h)
 
 namespace wil
 {
-    //! The key name includes the absolute path of the key in the registry, always starting at a
-    //! base key, for example, HKEY_LOCAL_MACHINE.
-    size_t const max_registry_key_name_length = 255;
+    namespace reg
+    {
+        /**
+         * \brief Helper function to translate registry return values if the value was not found
+         * \param hr HRESULT to test from registry APIs
+         * \return boolean if the HRESULT indicates the registry value was not found
+         */
+        constexpr bool is_hresult_not_found(HRESULT hr) WI_NOEXCEPT
+        {
+            return hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);

```

Comment:
Registry APIs can also yield ERROR_PATH_NOT_FOUND. I forget the exact cases (`RegOpenKeyEx(HKLM, "SOFTWARE\foo\bar")` when `foo` doesn't exist?) but we've got checks for it in our code due to bloody scars earned the hard way.

```
return (hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)) ||
        (hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));

Reply:
I added more tests for that, but couldn't reproduce that error. But I added it to the inline function anyway - just in case there are scenarios that I haven't touched.

Reply:
I think this is a worthy change, but I am worried about not being able to reproduce the issue.

[RegGetValue](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-reggetvaluea#return-value) claims it does not return `ERROR_PATH_NOT_FOUND` (by omission).

[RegCreateKey](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regcreatekeyexa#return-value) could possibly return it...

@DrusTheAxe, would you be able to follow up offline with the locations of some of these checks?